### PR TITLE
feat(core): more read options for ngIvy queries

### PR DESCRIPTION
**Pull request by pkozlowski-opensource**
More fun with the `read` option for queries. TBH I didn't know if / how all the cases described in the test work today... 

Unfortunately I've got a longish list of more corner cases I can think of, so stay tuned for more PRs like this...

### CODE REVIEW

File: packages/core/src/render3/instructions.ts [hunk 1]
```diff
@@ -840,6 +842,11 @@ export function directive<T>(index: number, directive?: T, directiveDef?: Direct
     ngDevMode && assertDataInRange(index - 1);
     Object.defineProperty(
         directive, NG_HOST_SYMBOL, {enumerable: false, value: previousOrParentNode});
+
+    if (localName != null) {
+      Object.defineProperty(directive, NG_LOCALNAME_SYMBOL, {enumerable: false, value: localName});
```
Comment (mhevery) [inline packages/core/src/render3/instructions.ts]: this is a slow operation and one which is megamorphic. We need to solve this in a way which does not require monkey patching.

File: packages/core/src/render3/query.ts [hunk 2]
```diff
@@ -133,6 +137,20 @@ function readFromNodeInjector(nodeInjector: LNodeInjector, node: LNode, read: Qu
...
 }
 
+function getDirectiveMatchingSelector(node: LNode, selector: string): any|null {
+  const ngData = node.view.data;
+  const flags = node.flags;
+  for (let i = flags >> LNodeFlags.INDX_SHIFT,
+           ii = i + ((flags & LNodeFlags.SIZE_MASK) >> LNodeFlags.SIZE_SHIFT);
+       i < ii; i++) {
+    const instance = ngData[i];
+    if (instance[NG_LOCALNAME_SYMBOL] === selector) {
```
Comment (mhevery) [inline packages/core/src/render3/query.ts]: this is a megamorphic read, hence slow. Need to solve it differently. We probably need to store it on the staticData

**Review by mhevery — CHANGES_REQUESTED** [review]

**Comment by pkozlowski-opensource** [comment]
@mhevery updated to use a different storage option as described in https://github.com/mhevery/angular/pull/21/files#diff-5505b78967166511d61d5d2acbbe454aR60

PTAL

**Review by mhevery — APPROVED** [review]

**Comment by pkozlowski-opensource** [comment]
@mhevery rebased, should be good to go in.

**Review by mhevery — APPROVED** [review]

**Comment by mhevery** [comment]
closing in favor of https://github.com/angular/angular/pull/21187
